Sources/PersistentCollections/PersistentDictionary._Node.swift

```diff
@@ -313,548 +310,473 @@ extension PersistentDictionary._Node {
     }
 
-    guard (trieMap & bitpos) == 0 else {
-      let index = _indexFrom(trieMap, mask, bitpos)
-      return self.getNode(index).get(key, keyHash, shift + _bitPartitionSize)
+    if trieMap.contains(bucket) {
+      let offset = trieMap.offset(of: bucket)
+      return self.getNode(offset).get(key, path.descend())
     }
 
     return nil
   }
 
-  func containsKey(_ key: Key, _ keyHash: Int, _ shift: Int) -> Bool {
-    let mask = _maskFrom(keyHash, shift)
-    let bitpos = _bitposFrom(mask)
-
-    guard collisionFree else {
-      let content: [ReturnPayload] = Array(self)
-      let hash = _computeHash(content.first!.key)
-
-      guard keyHash == hash else {
-        return false
-      }
-
+  func containsKey(_ key: Key, _ path: _HashPath) -> Bool {
+    guard isRegularNode else {
+      let content: [Element] = Array(self)
+      let hash = _HashValue(content.first!.key)
+      guard path._hash == hash else { return false }
       return content.contains(where: { key == $0.key })
     }
 
-    guard (dataMap & bitpos) == 0 else {
-      let index = _indexFrom(dataMap, mask, bitpos)
-      let payload = self.getPayload(index)
+    let bucket = path.currentBucket
+
+    if dataMap.contains(bucket) {
+      let offset = dataMap.offset(of: bucket)
+      let payload = self.getPayload(offset)
       return key == payload.key
     }
 
-    guard (trieMap & bitpos) == 0 else {
-      let index = _indexFrom(trieMap, mask, bitpos)
+    if trieMap.contains(bucket) {
+      let offset = trieMap.offset(of: bucket)
       return self
-        .getNode(index)
-        .containsKey(key, keyHash, shift + _bitPartitionSize)
+        .getNode(offset)
+        .containsKey(key, path.descend())
     }
 
     return false
   }
 
   func index(
-    _ key: Key,
-    _ keyHash: Int,
-    _ shift: Int,
+    forKey key: Key,
+    _ path: _HashPath,
     _ skippedBefore: Int
   ) -> Index? {
-    guard collisionFree else {
-      let content: [ReturnPayload] = Array(self)
-      let hash = _computeHash(content.first!.key)
-
-      assert(keyHash == hash)
+    guard isRegularNode else {
+      let content: [Element] = Array(self)
+      let hash = _HashValue(content.first!.key)
+      assert(path._hash == hash)
       return content
         .firstIndex(where: { _key, _ in _key == key })
         .map { Index(_value: $0) }
     }
 
-    let mask = _maskFrom(keyHash, shift)
-    let bitpos = _bitposFrom(mask)
-
-    let skipped = self._counts.prefix(upTo: mask).reduce(0, +)
+    let bucket = path.currentBucket
 
-    guard (dataMap & bitpos) == 0 else {
-      let index = _indexFrom(dataMap, mask, bitpos)
-      let payload = self.getPayload(index)
+    if dataMap.contains(bucket) {
+      let offset = dataMap.offset(of: bucket)
+      let payload = self.getPayload(offset)
       guard key == payload.key else { return nil }
-
-      return Index(_value: skippedBefore + skipped)
+      return Index(_value: skippedBefore + _count(upTo: bucket))
```

Comment:
Ah, one functional change slipped in -- calculating the subtree count on the fly here boosts some indexing operations.

We will not preserve the current `Index` representation, but these results will probably be representative of the eventual `distance(from:to:)` and `index(_:offsetBy:)` performance.

<img width="1280" alt="01 PersistentDictionary Int, Int  unsuccessful index(forKey:)" src="https://user-images.githubusercontent.com/608696/189416300-77984aed-981c-4cac-9e46-ee9831c07386.png">
<img width="1280" alt="02 PersistentDictionary Int, Int  successful index(forKey:)" src="https://user-images.githubusercontent.com/608696/189416303-4d754960-26ce-45ef-8c06-15e4fa0c99c3.png">
<img width="1280" alt="03 PersistentDictionary Int, Int  sequential iteration, indices" src="https://user-images.githubusercontent.com/608696/189416307-5fb8e9c1-0953-4286-ba67-0dacb00d4b89.png">
<img width="1280" alt="04 PersistentDictionary Int, Int  indexing subscript" src="https://user-images.githubusercontent.com/608696/189416308-9d1a47dc-3646-485e-af54-4821c968458f.png">